src/intermediate_config.rs

```diff
@@ -0,0 +1,330 @@
+        Yaml::Hash(remote_machine) => {
+            let host = match &remote_machine.get(&Yaml::String(String::from("host"))) {
+                Some(host) => match host {
+                    Yaml::String(host) => Some(host.to_string()),
+                    Yaml::Null => None,
+                    _ => return Err(String::from("remote.host must be a string"))
+                },
+                None => None
+            };
+
+            Some(IntermediateRemote {
+                host,
+            })
+        }
+        Yaml::Null | Yaml::BadValue => None,
+        ref something_else => return Err(format!("'remote' must be an object, but was {:#?}", something_else))
+    };
+
+    let push = match &yaml["push"] {
+        Yaml::Hash(push) => {
+            let compression = parse_compression(push, "compression", "push");
+
+            match compression {
+                Ok(value) => Some(IntermediatePush {
+                    compression: value
+                }),
+                Err(error) => return Err(error)
+            }
+        }
+        Yaml::Null | Yaml::BadValue => None,
+        _ => return Err(String::from("'push' must be an object"))
+    };
+
+    let pull = match &yaml["pull"] {
+        Yaml::Hash(pull) => {
+            let compression = parse_compression(pull, "compression", "pull");
+
+            match compression {
+                Ok(value) => Some(IntermediatePull {
+                    compression: value,
+                }),
+                Err(error) => return Err(error)
+            }
+        }
+        Yaml::Null | Yaml::BadValue => None,
+        _ => return Err(String::from("'pull' must be an object"))
+    };
+
+    Ok(IntermediateConfig {
+        remote,
+        push,
+        pull,
+    })
+}
+
+fn parse_compression(yaml: &LinkedHashMap<Yaml, Yaml>, field_name: &str, scope_name: &str) -> Result<Option<u8>, String> {
+    match yaml.get(&Yaml::String(field_name.to_string())).cloned() {
+        Some(compression) => match compression {
+            Yaml::Integer(compression) => if compression >= 1 && compression <= 9 {
+                Ok(Some(compression as u8))
+            } else {
+                Err(format!("'{}.{}' must be a positive integer from 1 to 9, but was {:#?}", scope_name, field_name, compression))
+            },
+            Yaml::Null | Yaml::BadValue => Ok(None),
+            ref something_else => Err(format!("'{}.{}' must be a positive integer from 1 to 9, but was {:#?}", scope_name, field_name, something_else))
+        },
+        None => Ok(None)
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn parse_config_from_str_all_fields_2_spaces_indent() {
+        let content = "
+remote:
+  host: computer1
+push:
+  compression: 5
+pull:
+  compression: 2
+";
+
+        assert_eq!(parse_config_from_str(content), Ok(IntermediateConfig {
+            remote: Some(IntermediateRemote {
+                host: Some(String::from("computer1")),
```

Comment:
Isn’t it better to define expected variables beforehand and put them into the `content`?